File: src/BaekJoon/Java/baekjoon_1987.java
package BaekJoon.Java;
// https://www.acmicpc.net/problem/1987

import java.util.Scanner;

public class baekjoon_1987 {
	static int R, C, max = 0;
	static char[][] arr;
	static boolean[] v = new boolean[26];
	static int[] dx = {-1, 0, 1, 0};
	static int[] dy = {0, 1, 0, -1};

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		R = sc.nextInt();
		C = sc.nextInt();
		arr = new char[R][C];
		sc.nextLine();

		for(int i = 0; i < R; i++)
			arr[i] = sc.nextLine().toCharArray();

		v[arr[0][0]-65] = true;
		dfs(0, 0, 1);
		System.out.println(max);
		sc.close();
	}

	public static void dfs(int x, int y, int cnt) {
		if(cnt > max)
			max = cnt;

		for(int i = 0; i < 4; i++) {
			int currX = x + dx[i];
			int currY = y + dy[i];

			if(currX >= 0 && currY >= 0 && currX < R && currY < C) {
				int here = arr[currX][currY] - 65;

				if(!v[here]) {
					v[here] = true;
					dfs(currX, currY, cnt+1);
					v[here] = false;
				}
			}
		}
	}
}

File: src/BaekJoon/Rust/baekjoon_1987.rs
use std::io::stdin;

static mut MAX: i32 = 0;
static DX: [i32; 4] = [-1, 0, 1, 0];
static DY: [i32; 4] = [0, 1, 0, -1];

pub fn baekjoon_1987() {
    let r_c: Vec<usize> = stdin().lines().next().unwrap().unwrap()
        .split_whitespace().map(|s| s.parse::<usize>().unwrap()).collect::<Vec<usize>>();
    let (r, c) = (r_c[0], r_c[1]);
    let mut field: Vec<Vec<char>> = Vec::with_capacity(r);
    let mut visited: Vec<bool> = vec![false; 26];
    
    for i in 0..r {
        let mut line = String::new();
        
        stdin().read_line(&mut line).unwrap();
        
        field.push(line.trim().chars().collect());
    }
    
    unsafe {
        visited[(field[0][0] as u8 - b'A') as usize] = true;
        dfs(&field, &mut visited, 0, 0, 1);
        println!("{}", MAX);
    }
}

unsafe fn dfs(field: &Vec<Vec<char>>, visited: &mut Vec<bool>, x: i32, y: i32, count: i32) {
    if count > MAX {
        MAX = count;
    }
    
    for i in 0..4 {
        let curr_x: i32 = x + DX[i];
        let curr_y: i32 = y + DY[i];
        
        if curr_x >= 0 && curr_y >= 0 && (curr_x as usize) < field.len() && (curr_y as usize) < field[0].len() {
            let here: usize = (field[curr_x as usize][curr_y as usize] as u8 - b'A') as usize;
            
            if !visited[here] {
                visited[here] = true;
                dfs(field, visited, curr_x, curr_y, count + 1);
                visited[here] = false;
            }
        }
    }
}